04-onion-routing.md

```diff
@@ -261,6 +268,197 @@ The reader:
+
+One key distinction with `option_amp` is that the sender generates all
+preimages, and only reveals them to the final hop if all partial payments arrive
+successfully. As such, the payer _will not_ learn a new preimage as it would in
+the regular payment flow. For accounting purposes, however, `option_amp` can be
+used to fulfill an invoice akin to the regular payment flow, and also enforce
+additional constraints such as amounts and timelocks.
+
+The writer:
+ - MUST NOT include `option_amp` for any non-final node.
+ - if the sender has an invoice and `option_amp` feature was not set in the invoice:
+   - MUST NOT include `option_amp` for the final node.
+ - otherwise:
+   - MAY include `option_amp` for the final node.
+   - if it does include `option_amp`:
+     - MUST generate a random `stream_id` to be used on all HTLCs in the set.
+     - MAY send more than one HTLC using the same `stream_id`.
+     - MUST set the `share` values of all HTLCs such that their xor is a random
+       root seed `r`.
+     - SHOULD choose a unique child_index_i for each HTLC.
+     - MUST derive the `payment_hash` for an HTLC using `amp_child(r, child_index_i)`.
+     - if the invoice specifies a non-zero `amount`:
+       - MUST set `total_msat` to `amount`.
+     - otherwise:
+       - MUST set `total_msat` to the amount it wishes to pay.
+     - MUST ensure the total `amount_to_forward` in the HTLC set which arrives at
+       the payee is equal to `total_msat`.
+     - if the sender has an invoice:
+       - MUST set the `payment_id` of each HTLC to the `payment_hash` in the
+         invoice.
+     - otherwise:
+       - MUST set the `payment_id` of each HTLC to zero.
+
+The reader:
+ - if `option_amp` is present:
+  - MUST fail the HTLC if it is not the final node.
+  - MUST fail the HTLC as it would otherwise fail a single HTLC of
+    `amt_to_forward`, `payment_hash`, and `cltv_expiry` without context of the
+    invoice.
+  - if the `payment_id` is non-zero:
+    - MUST fail the HTLC if an invoice for `payment_id` does not exist.
+    - MUST fail the HTLC if `total_msat` is less than the invoice's `amount`.
+    - MUST fail the HTLC if `cltv_expiry` does not satisfy the invoice's
+      `min_final_cltv_expiry`.
+  - otherwise:
+    - MUST fail the HTLC if `cltv_expiry` does not satisfy the node's default
+      `min_final_cltv_expiry`
+  - MUST fail the entire HTLC set if `total_msat` is not the same for all HTLCs
+    in the set.
+  - if the total `amount_to_forward` of the HTLC set is equal `total_msat`:
+    - MUST reconstructs `r` as the xor of all `share`s in the HTLC set.
+    - MUST compute `p_i, h_i = amp_child(r, child_index_i)` for all HTLCs in the
+      set.
+    - if any `i-th` HTLC's `payment_hash` differs from `h_i`:
+      - MUST fail the HTLC set.
+    - otherwise:
+      - MAY fulfill the `i-th` HTLC in the set using `p_i`.
+  - otherwise:
+    - MUST fail an HTLC in set if its `cltv_expiry` elapses.
```

Comment:
Shouldn't it be failed earlier than that? In case of an (amp) hodl invoice, an application assumes that when the invoice is marked as accepted (we know the root seed, but haven't pulled yet), the invoice cltv expiry condition holds.

Reply:
Example:
* current height: 100
* (hodl) invoice cltv delta: 40, amt: 100
* first htlc for 50 sats comes in with expiry 140 -> accepted as partial payment
* new block comes in
* second htlc for 50 sats comes in with expiry 141 -> accepted as partial payment

The hodl invoice should now move to the accepted state, because enough has been paid. We'd send the `accepted` event to rpc subscribers. But at that point, there are only 39 blocks left before the first htlc expires. The subscriber will probably assume that when the event comes in, the final cltv delta of 40 is still met.